src/QsCompiler/CompilerOptimizations/ConstantPropagation.fs

```diff
@@ -0,0 +1,74 @@
+namespace Microsoft.Quantum.QsCompiler.CompilerOptimization
+
+open System.Collections.Immutable
+open Microsoft.Quantum.QsCompiler.SyntaxExtensions
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+
+open Microsoft.Quantum.QsCompiler.CompilerOptimization.Utils
+open Microsoft.Quantum.QsCompiler.CompilerOptimization.ExpressionEvaluation
+
+
+/// The SyntaxTreeTransformation used to evaluate constants
+type ConstantPropagator(compiledCallables: ImmutableDictionary<QsQualifiedName, QsCallable>) =
+    inherit SyntaxTreeTransformation()
+    let vars = VariablesDict()
+    let cd = {compiledCallables = compiledCallables}
+
+    let mutable changed = true
+    let mutable prevChanged = false
+    let mutable declarations = []
+
+    /// Returns whether the syntax tree has been modified since this function was last called
+    member this.checkChanged() =
+          let x = changed
+          changed <- false
+          x
+
+    /// Marks the syntax tree as having changed
+    member this.markChanged() =
+        prevChanged <- changed
+        changed <- true
+
+    /// Ignores the most recent marking of the syntax tree changing
+    member this.undoMarkChanged() =
+        changed <- prevChanged
+
+    /// Gets a sorted list of the names of all the constant local variables
+    member this.getDeclarations =
```

Comment:
Unless I am overlooking something, it seems that this method is mostly used for logging/testing purposes? 
If so, please make it internal, and add the corresponding make internals visible to attribute to the project for testing purposes if needed.